### PR TITLE
harden SSZ reader

### PR DESCRIPTION
**Pull request by tersec**
Addresses https://github.com/status-im/nim-beacon-chain/issues/896

There are probably other missing validation checks, but this is a relevant one that addresses that fuzzing SSZ  (and other invalid inputs). The whole variable-length reading aspect is a bit too trusting at the moment.

### CODE REVIEW

**Comment by arnetheduck** [comment]
test case?

**Comment by tersec** [comment]
https://github.com/status-im/nim-beacon-chain/files/4487311/indexError_beaconstate_empty_container_nimbus_devel.zip

**Comment by tersec** [comment]
There's a challenge with SSZ-testcases, though, as they'll all stop working if there's ever another breaking change in phase0 `BeaconState`.

**Comment by tersec** [comment]
The `status-im.nim-beacon-chain (Windows 32-bit)` failure is:
```
Starting: build and test
==============================================================================
Task         : Bash
Description  : Run a Bash script on macOS, Linux, or Windows
Version      : 3.163.1
Author       : Microsoft Corporation
Help         : https://docs.microsoft.com/azure/devops/pipelines/tasks/utility/bash
==============================================================================
Generating script.
"C:\Program Files\Git\bin\bash.exe" --noprofile --norc -c pwd
/d/a/_temp

========================== Starting Command Output ===========================
"C:\Program Files\Git\bin\bash.exe" --noprofile --norc /d/a/_temp/ec56aa29-9e9b-46a1-931a-8bd889a59444.sh
Installing MinGW-w64
Fetching submodules
Downloading official test vectors

##[error]Bash exited with code '1'.
Finishing: build and test
```

File: beacon_chain/ssz/bytes_reader.nim [hunk 1]
```diff
@@ -169,6 +169,8 @@ func readSszValue*(input: openarray[byte], T: type): T =
           raise newException(MalformedSszError, "SSZ field offsets are not monotonically increasing")
         elif endOffset > input.len:
           raise newException(MalformedSszError, "SSZ field offset points past the end of the input")
+        elif startOffset < boundingOffsets[0]:
```
Comment (zah) [inline beacon_chain/ssz/bytes_reader.nim:172]: This check doesn't look right though. `boundingOffsets[0]` here will be the fixed position where the offset of the dynamic field appears. All such offsets are part of the object which is known in the code as `fixedPortionSize`. 

The problem here is that the `startOffset` cannot point to any position within this `fixedPortionSize`, so the check must be `startOffset < minimallyExpectedSize` (this is a constant defined 10 lines above and equal to the `fixedPortionSize`)

Comment (tersec) [reply beacon_chain/ssz/bytes_reader.nim]: https://github.com/status-im/nim-beacon-chain/pull/898/commits/07328bf552a015248fe7c4d540530c7acea0aa82

**Pull request merged**